core/src/main/scala/org/apache/spark/deploy/worker/DriverRunner.scala

```diff
@@ -68,56 +71,67 @@ private[deploy] class DriverRunner(
 
   private var clock: Clock = new SystemClock()
   private var sleeper = new Sleeper {
-    def sleep(seconds: Int): Unit = (0 until seconds).takeWhile(f => {Thread.sleep(1000); !killed})
+    def sleep(seconds: Int): Unit = Thread.sleep(seconds * 1000)
   }
 
   /** Starts a thread to run and manage the driver. */
   private[worker] def start() = {
-    new Thread("DriverRunner for " + driverId) {
+    workerThread = new Thread("DriverRunner for " + driverId) {
       override def run() {
         try {
-          val driverDir = createWorkingDirectory()
-          val localJarFilename = downloadUserJar(driverDir)
-
-          def substituteVariables(argument: String): String = argument match {
-            case "{{WORKER_URL}}" => workerUrl
-            case "{{USER_JAR}}" => localJarFilename
-            case other => other
+          shutdownHook = ShutdownHookManager.addShutdownHook { () =>
+            killProcessAndFinalize(DriverState.KILLED, new SparkException("Worker shutting down"))
```

Comment:
There is a race condition here: if this shutdown hook is running while the thread is starting the driver process, the shutdown hook may miss that.